README.md
# episheet

R functions for Modern Epidemiology.

## CRAN details

<table>
  <tr>
    <td><strong>Link to CRAN entry</strong></td>
    <td><a href="http://cran.r-project.org/package=episheet"><strong>http://cran.r-project.org/package=episheet</strong></a></td>
  </tr>
  <tr>
    <td><strong>Unix build status of last commit</strong><p><sup>Passing if devtools::check() error free</sup></p></td>
    <td><img src="https://travis-ci.org/epijim/episheet.svg?branch=master" alt="Badge"></td>
  </tr>
  <tr>
    <td><strong>Current version on CRAN</strong><p><sup>This repo may be ahead of CRAN</sup></td>
    <td><img src="http://www.r-pkg.org/badges/version/episheet" alt="Badge"></td>
  </tr>
  <tr>
    <td><strong>Total downloads from CRAN</strong><p><sup>Because....vanity</sup></p></td>
    <td><img src="http://cranlogs.r-pkg.org/badges/grand-total/episheet" alt="Badge"></td>
  </tr>
   <tr>
    <td><strong>Documentation</strong><p><sup>Click badge to see.</sup></p></td>
    <td><a href="http://www.rdocumentation.org/packages/episheet">
<img src="http://www.rdocumentation.org/badges/version/episheet" alt="http://www.rdocumentation.org/packages/episheet"></td>
    </a>
  </tr>
</table>

## Text book

This package is functions to support the text book

> Modern Epidemiology, Second Edition, by Kenneth J. Rothman and Sander Greenland.

Textbook website: [krothman.org](http://krothman.org/)

## Functions

**Note** - only a few functions have been mapped to R so far!

* Rate Data	(mapped)			
* Case-control Power				
* Matched Case-control data						
* Risk Data	(mapped)							
* Study Size				
* O. R. Exact						
* Case-control Data				
* Quickcalc				
* P-Value Functions
    + `pvalueplot` Plot a p-value function
    + [Example of use](https://cran.r-project.org/web/packages/episheet/vignettes/pvalueplot.html)
* Standardization				
* Seasonal Analysis				
* Meta-Analysis (contributed by T. Andersson & A. Ahlbom)						
* Cohort Power				
* Life Table										

## Complete excel tool

This package currently contains functions for calculating rate ratios and differences, risk ratios and differences, Mantel Haenzsel risk ratio. The complete set is available via the excel tool [episheet](http://krothman.org/Episheet.xls).


NEWS.md

# episheet 0.3.9999

Added function to calculate rate ratios and rate differences. 
Also fixed bug in which confidence intervals for risk ratios and risk differences did not calculate correctly. 

Added example cohort data on ebola survival. 

# episheet 0.3.0

Simon added functions for calculating risk ratios, risk differences and MH risk ratios for stratified risks. 

# episheet 0.2.0

Based on feedback after CRAN release, including additional feedback from Rothman, the option set has been expanded on the function pvalueplot to make it easier to build publication quality plots.

# episheet 0.1.0

First release. One function added.


tests/testthat.R
library(testthat)
library(episheet)

test_check("episheet")


tests/testthat/test-risk.R
context("Test risk")

# dat <- data.frame(
#     exposure_var = c(rep(1, 8), rep(0, 5), rep(1, 98), rep(0, 115)),
#   outcome_var = c(rep(1, 8), rep(1, 5), rep(0, 98), rep(0, 115)),
#   stringsAsFactors = FALSE
# )

# p 248 table 14-4 for check
dat <- data.frame(
  exposure_var = c(rep(0, 7), rep(1, 12), rep(0, 9), rep(1, 2)),
  outcome_var = c(rep(1, 7), rep(1, 12), rep(0,9), rep(0, 2))
)

test_that("Test risk_ratio returns exected estimate", {
  expect_equal(risk(data = dat, exposure = exposure_var, outcome = outcome_var
                    )$risk_ratio[[2]], 1.96, tol = 0.01)
})

test_that("Test risk_ratio returns exected lci", {
  expect_equal(risk(data = dat, exposure = exposure_var, outcome = outcome_var
  )$rr_lci[[2]], 1.1, tol = 0.1)
})

test_that("Test risk_ratio returns exected uci", {
  expect_equal(risk(data = dat, exposure = exposure_var, outcome = outcome_var
  )$rr_uci[[2]], 3.6, tol = 0.1)
})

test_that("Test risk_diff returns exected estimate", {
  expect_equal(risk(data = dat, exposure = exposure_var, outcome = outcome_var
  )$risk_diff[[2]], 42, tol = 0.001)
})

test_that("Test risk_diff returns exected lci", {
  expect_equal(risk(data = dat, exposure = exposure_var, outcome = outcome_var
  )$rd_lci[[2]], 10, tol = 0.1)
})

test_that("Test risk_diff returns exected uci", {
  expect_equal(risk(data = dat, exposure = exposure_var, outcome = outcome_var
  )$rd_uci[[2]], 73, tol = 0.1)
})


R/data.R
#' tolbutamide data - chapter 15-1, p260
#'
#' Age-specific comparison of death from all causes for Tolbutamide and placebo
#' treatment groups, University Group Diabetes Program (1970)
#'
#'  @format A dataframe with 409 observations and 3 variables
#'  \describe{
#'    \item{tolbutamide}{Given Tolbutamide (1) or placebo (0)}
#'    \item{dead}{Died (1) or surviving (0)}
#'    \item{age}{Less than 55 (<55) or 55 and over (ge55)}
#'  }
"tolbutamide"

#' ebola data
#'
#' Person-time data for death from ebola. Data from
#' https://datacompass.lshtm.ac.uk/599/
#'  Marks, M. Learning Clinical Epidemiology with R. (Project). London School of Hygiene & Tropical Medicine, London, United Kingdom.
#'
#' @format A dataframe
#' \describe{
#'    \item{id}{Unique patient identifier}
#'    \item{age}{Age in years}
#'    \item{age_group}{Grouped age}
#'    \item{sex}{Sex}
#'    \item{disease_onset}{Date of onset (date class)}
#'    \item{disease_ended}{Date of death or recovery (date class)}
#'    \item{days_at_risk}{Number of days at risk of death}
#'    \item{status}{Died or recovered (string)}
#'    \item{transmission}{How ebola was transmitted}
#'    \item{died}{Integer 1 = died, 0 = survived}
#' }
"ebola"


R/rate.R
#' Calculate risk ratio and risk difference
#'
#' Calculate risk ratios and risk differences using a Poisson distribution for
#' person-time data.
#' Function works on individual level data or aggregated data
#' p244 2nd Edition
#'
#' @param data A dataframe
#' @param outcome Variable with the outcomes as a numeric variable
#' @param denominator Variable giving the amount of time at risk
#' @param exposure Variable giving whether exposed or not
#' @param per_unit Multiplier for rate values, e.g. \code{1000} for n outcomes per 1000 denominator
#' @param ci_level A numeric value giving the confidence interval
#'
#' @export
#'
#' @examples
#'
#' # Using individual level data
#'
#' data(ebola)
#' library(dplyr)
#' ebola %>%
#'  mutate(male = ifelse(sex == "male", 1, 0)) %>%
#'   rate(outcome = died, denominator = days_at_risk, exposure = male,
#'   per_unit = 100)
#'
#' # Using aggregated data
#' # Table 14-2
#' cancer_xray <- data.frame(cases = c(41, 15), pyar = c(28010, 19017),
#' radiation = c(1, 0))
#' cancer_xray
#' cancer_xray %>%
#'   rate(outcome = cases, denominator = pyar, exposure = radiation,
#'     per = 100000)

rate <- function(data, outcome, denominator, exposure, per_unit,
                 ci_level = 95){

  #Fixing undefined global variable issue from CRAN check.
  # hacky though
  base_rate <- rate_diff <- rate_ratio <- NULL

  exposure <- dplyr::enquo(exposure)
  outcome <- dplyr::enquo(outcome)
  denominator <- dplyr::enquo(denominator)

  assertthat::assert_that(dplyr::between(ci_level, 0.1, 99.9),
                          msg = "ci_level must be between 0.1 and 99.9")


    z <- abs(qnorm((100 - 95)/2/100))

    data <- data %>% dplyr::filter(!is.na(!!outcome) & !is.na(!!denominator) &
                            !is.na(!!exposure))

    A0 <- data %>% dplyr::group_by(!!exposure) %>%
      dplyr::summarise(outcome = sum(!!outcome)) %>%
      dplyr::ungroup() %>%
      dplyr::filter(!!exposure == 0) %>%
      dplyr::pull()

    A1 <- data %>% dplyr::group_by(!!exposure) %>%
      dplyr::summarise(outcome = sum(!!outcome)) %>%
      dplyr::ungroup() %>%
      dplyr::filter(!!exposure == 1) %>%
      dplyr::pull()

    sd_ln_rr <- (1/A0 + 1/A1) ^ (1/2)

    T0 <- data %>% dplyr::group_by(!!exposure) %>%
      dplyr::summarise(outcome = sum(!!denominator, na.rm = TRUE)) %>%
      dplyr::ungroup() %>%
      dplyr::filter(!!exposure == 0) %>%
      dplyr::pull()

    T1 <- data %>% dplyr::group_by(!!exposure) %>%
      dplyr::summarise(outcome = sum(!!denominator, na.rm = TRUE)) %>%
      dplyr::ungroup() %>%
      dplyr::filter(!!exposure == 1) %>%
      dplyr::pull()

    sd_ln_rd <- ((A1/(T1)^2 + A0/(T0)^2) ^ (1/2)) * per_unit

  z <- data %>%
    dplyr::group_by(!!exposure) %>%
    dplyr::summarise(outcome = sum(!!outcome),
                     denominator = sum(!!denominator)) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(rate = (outcome / denominator) * per_unit,
                  base_rate = rate[dplyr::row_number(rate) == 1],
                  rate_ratio = rate / base_rate,
                  rate_ratio_lci = exp(log(rate_ratio) - (z * sd_ln_rr)),
                  rate_ratio_uci = exp(log(rate_ratio) + (z * sd_ln_rr)),
                  rate_diff = rate - base_rate,
                  rate_diff_lci = rate_diff - z * sd_ln_rd,
                  rate_diff_uci = rate_diff + z * sd_ln_rd
                  ) %>%
    dplyr::select(-base_rate)

  return(z)

}


R/risk.R
#' Calculate risk ratio and risk difference
#'
#' Calculate risk ratios and risk differences
#'
#' @param data A dataframe
#' @param exposure Variable giving the levels of the outcome
#' @param outcome Variable giving cases (1) or non-cases (0)
#' @param ci_level a string giving the confidence interval
#'
#' @examples
#' # Data from stratum 1 of table 15-1., p260
#' dat <- data.frame(
#'     exposure_var = c(rep(1, 8), rep(0, 5), rep(1, 98), rep(0, 115)),
#'   outcome_var = c(rep(1, 8), rep(1, 5), rep(0, 98), rep(0, 115)),
#'   stringsAsFactors = FALSE
#' )
#' risk(data = dat, exposure = exposure_var, outcome = outcome_var)
#' @return A dataframe with the risk ratio, risk difference and confidence intervals
#' @export

risk <- function(data, exposure, outcome, ci_level = 95){

  #Fixing undefined global variable issue from CRAN check.
  # hacky though
  n_var <- base_risk <- risk_ratio <- risk_diff <- n<- NULL

  exposure <- dplyr::enquo(exposure)
  outcome <- dplyr::enquo(outcome)

  assertthat::assert_that(dplyr::between(ci_level, 0.1, 100),
                          msg = "ci_level must be between 0.1 and 100")

  assertthat::assert_that(all(data[,rlang::quo_name(outcome)] %in% c(0,1)) ,
                          msg = "Outcome must be either 1 or 0")

  assertthat::assert_that(all(data[, rlang::quo_name(exposure)] %in% c(0,1)),
                          msg = "Exposure must be either 1 or 0")

  # Notation:
    # E: Expected nuber of cases
    # V: variance
    # A1: exposed cases, A2 Unexposed cases, B1 exposed non-cases B2 unexposed non-cases
    # N1 total exposed, N2 total unexposed, M1 total cases, M2 total non-cases N total observations

  # p 248 for formulae
  A1 <- data %>% dplyr::filter(!!exposure == 1 & !!outcome == 1) %>%
    dplyr::summarise(n = n()) %>% dplyr::pull()
  A0 <- data %>% dplyr::filter(!!exposure == 0 & !!outcome == 1) %>%
    dplyr::summarise(n = n()) %>% dplyr::pull()
  B1 <- data %>% dplyr::filter(!!exposure == 1 & !!outcome == 0) %>%
    dplyr::summarise(n = max(dplyr::row_number(!!exposure))) %>% dplyr::pull()
  B0 <- data %>% dplyr::filter(!!exposure == 0 & !!outcome == 0) %>%
    dplyr::summarise(n = max(dplyr::row_number(!!exposure))) %>% dplyr::pull()

  # E = (M1 * N1) / N
  E <- (sum(data[,rlang::quo_name(outcome)]) * sum(data[, rlang::quo_name(exposure)])) /
    length(data[,rlang::quo_name(outcome)])
  # V = (E * M0 * N0) / (N * N-1)
  M0 <- length(data[,rlang::quo_name(outcome)]) - sum(data[,rlang::quo_name(outcome)])
  N0 <- length(data[, rlang::quo_name(exposure)]) - sum(data[, rlang::quo_name(exposure)])
  N1 <- data %>% dplyr::summarise(n1 = sum(!!exposure)) %>% dplyr::pull()
  V <- (E * M0 * N0)/(length(data[,rlang::quo_name(outcome)]) *
                        (length(data[,rlang::quo_name(outcome)])-1) )

  # sd_ln_rr = ( 1/A1 - 1/N1 + 1/A0 - 1/N0) ^ (1/2)
  sd_ln_rr <- (1/A1 - 1/sum(data[, rlang::quo_name(exposure)]) + 1/A0 - 1/N0 ) ^ (1/2)
  # sd_ln_rd = ( ((A1*B1)/(N1^2 * (N1 - 1))) + ( (A0*B0)/ (N0 ^ 2 * (N0 - 1) ) ) )^(1/2)
  sd_rd <-( ((A1*B1)/(N1^2 * (N1 - 1))) + ( (A0*B0)/ (N0 ^ 2 * (N0 - 1) ) ) ) ^ (1/2)
  # z <- abs(qnorm((100 - ci_level)/100))
  z <- abs(qnorm(((100 - 95)/2)/100))


  z <- data %>%
    dplyr::group_by(!!exposure) %>%
    dplyr::summarise(n = n(), outcome = sum(!!outcome)) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(risk = (outcome / n) * 100,
                  base_risk = risk[dplyr::row_number(risk) == 1],
                  risk_ratio = risk / base_risk,
                  rr_lci = exp(log(risk_ratio) - z * sd_ln_rr),
                  rr_uci = exp(log(risk_ratio) + z * sd_ln_rr),
                  risk_diff = risk - base_risk,
                  rd_lci = risk_diff - (z * sd_rd)*100,
                  rd_uci = risk_diff + (z * sd_rd)*100
                  ) %>%
    dplyr::select(-base_risk)
  return(z)
}


R/stratified_risk.R
#' Stratified risk
#'
#' Calculate stratified risk estimates as per Chapter 15,
#'
#' @param data A dataframe providing the exposure, outcome and stratifying variable
#' @param exposure binary variable giving the exposure status
#' @param outcome binary variable giving the outcome status
#' @param stratifier stratifying variable
#' @param ci_level variable giving the limits for the confidence interval
#'
#' @return A dataframe giving an MH-adjusted risk ratio
#' @export
#' @examples
#' data(tolbutamide)
#' stratified_risk(tolbutamide, exposure = tolbutamide, outcome = dead,
#'     stratifier = age)

stratified_risk <- function(data, exposure, outcome, stratifier, ci_level = 95){

  . <- NULL

  # Fixing CRAN check undefined global variable warning
  aN0T <-  base_risk <-  bN1T <-  exposed <-  lci <-  m1n1n0 <- rrmh <- NULL
  Tc <- Te <- Tec <- Ts <- Tu <- Tuc <- n <- uci <- var <- NULL
  exposure <- dplyr::enquo(exposure)
  outcome <- dplyr::enquo(outcome)
  stratifier <- dplyr::enquo(stratifier)

  assertthat::assert_that(dplyr::between(ci_level, 0.1, 100),
                          msg = "ci_level must be between 0.1 and 100")

  assertthat::assert_that(all(data[,rlang::quo_name(outcome)] %in% c(0,1)) ,
                          msg = "Outcome must be either 1 or 0")

  assertthat::assert_that(all(data[, rlang::quo_name(exposure)] %in% c(0,1)),
                          msg = "Exposure must be either 1 or 0")

  alpha <- ((100 - ci_level) / 100) / 2
  z <- abs(qnorm(alpha))

  # RRmh in sheet is AH18/AI18
  # AH18 is sum of aN0/T across strata
  #     (aN0/T is L2*M4/N4 - i.e. exposed cases * total unexposed over absolute total
  # AI18 is sum of bN1/T across strata
  #     (bN1/T is M2*L4/N4
  #     m2 is unexposed cases, L4 is total exposed, N4 is absolute total
  # ie, total exposed over total)

  totals <- data %>%
    dplyr::group_by(!!stratifier) %>%
    dplyr::summarise(Ts = n(), Tc = sum(!!outcome)) %>%
    dplyr::ungroup()

  exposed_totals <- data %>%
    dplyr::filter(!!exposure == 1) %>%
    dplyr::group_by(!!stratifier) %>%
    dplyr::summarise(Te = n(), Tec = sum(!!outcome)) %>%
    dplyr::ungroup()


  unexposed_totals <- data %>%
    dplyr::filter(!!exposure == 0) %>%
    dplyr::group_by(!!stratifier) %>%
    dplyr::summarise(Tu = n(), Tuc = sum(!!outcome)) %>%
    dplyr::ungroup()

  aN0 <- totals %>%
    dplyr::left_join(., exposed_totals) %>%
    dplyr::left_join(unexposed_totals) %>%
    dplyr::mutate(aN0T = Tec * Tu / Ts,
                  bN1T = Tuc * Te / Ts,
                  m1n1n0 = (Tc * Te * Tu) / Ts^2 - Tec * Tuc / Ts
    )

  # RRmh in sheet is AH18/AI18

  # Variance of RRMH is AJ18 / ah18 * ai18
  # aj is (n2*l4*m4) / n4^2 - l2 * M2 / N4
  # n2 is Tc, l4 is Te, m4 is Tu, n4 is Ts
  # l2 is Tec, m2 is Tuc, n4 is Ts


  z <- data %>%
    dplyr::group_by(!!stratifier) %>%
    dplyr::summarise(n = n(), exposed = sum(!!exposure), outcome = sum(!!outcome)) %>%
    dplyr::left_join(., aN0) %>%
    dplyr::ungroup() %>%
    dplyr::mutate(
      rrmh = sum(aN0T)/sum(bN1T),
      var = sum(m1n1n0) / (sum(aN0T) * sum(bN1T)),
      # sum_m1n1no = sum(m1n1n0),
      # sum_aN0T = sum(aN0T),
      # sum_bN1T = sum(bN1T),
      lci = exp(log(rrmh) - z * sqrt(var)),
      uci = exp(log(rrmh) + z * sqrt(var)),
    ) %>%
    dplyr::select(!!stratifier, n, exposed, outcome, rrmh, var, lci, uci)
  return(z)
}


R/pvalueplot.R
#' Plot the p-value function
#'
#' Plot the p-value function for one or two confidence interval pairs.
#' See following for example of the use in the literature:
#' Is flutamide effective in patients with bilateral orchiectomy?
#' Rothman, Kenneth J et al.
#' The Lancet , Volume 353 , Issue 9159 , 1184
#'
#' @section Bugs:
#' Code repo: \url{https://github.com/epijim/episheet}
#'
#' @param est1.ll Lower confidence interval of estimate 1
#' @param est1.ul Upper confidence interval of estimate 1
#' @param est2.ll Lower confidence interval of estimate 2 (optional)
#' @param est2.ul Upper confidence interval of estimate 2 (optional)
#' @param label1 If using two estimates, name the 1st
#' @param label2 If using two estimates, name the 2nd
#' @param xlabel The x axis label
#' @param citype Choose between '95\%CI', '90\%CI' or '99\%CI'
#' @param labelsize Change size of labels
#' @param functionwidth Change width of pvalue function line
#' @param referencewidth Change width of reference lines
#' @keywords R Rothman pvalues episheet
#' @export
#' @importFrom magrittr "%>%"
#' @importFrom graphics "curve"
#' @importFrom stats "qnorm"
#' @examples
#' pvalueplot(
#'   est1.ll = 0.9,
#'   est1.ul = 12,
#'   xlabel = "Relative Risk"
#' )
#'
#' pvalueplot(
#'   est1.ll = 0.8,
#'   est1.ul = 3.8,
#'   est2.ll = 1.2,
#'   est2.ul = 2,
#'   label1 = "Estimate 1",
#'   label2 = "Estimate 2",
#'   xlabel = "Relative Risk",
#'   citype = "95%CI"
#' )

pvalueplot <- function(
  est1.ll,
  est1.ul,
  est2.ll = NA,
  est2.ul = NA,
  label1 = "Estimate 1",
  label2 = "Estimate 2",
  xlabel = "Relative Risk",
  citype = "95%CI",
  labelsize = NULL,
  functionwidth = 1,
  referencewidth = 1
){
  # fake global variables to pass CMD check
  pvalue <- NULL
  updown <- NULL
  xvalue <- NULL
  zvalue <- NULL

  # which ci type?

  citype.pow <- ifelse(
    citype == "90%CI", 3.29,
    ifelse(citype == "99%CI", 5.16,
           3.84)
  )

  citype.pval <- ifelse(
    citype == "90%CI", 0.1,
    ifelse(citype == "99%CI", 0.01,
           0.05)
  )

  message(paste("Confidence interval type is ",citype))

  # calculate relative risk
  # Gardner M J and Altman D G. Statisitics with confidence.
  # BMJ publications. Reprint 1994 p 51-52
  jb_getrr <- function(ll,ul){
    exp((log(ll) + log(ul))/2)
  }

  # calculate standard errors
  jb_getse <- function(ll,ul,citype.pow){
    (log(ul) - log(ll))/citype.pow
  }

  # calculate curve value
  # up
  jb_getcurveup <- function(rr,se,pvalue){
    exp(log(rr) - pvalue*(se))
  }
  # down
  jb_getcurvedown <- function(rr,se,pvalue){
    exp(log(rr) + pvalue*(se))
  }

  # p-values of interest
  d.frame <- data.frame(
    # pvalues to plot
    pvalue = c(seq(0.01,1,0.01),seq(0.99,0.01,-0.01)),
    # formula changes if going up or down x values
    updown = c(rep("up",100),rep("down",99))
  ) %>%
    dplyr::mutate(
      zvalue = qnorm(1 - pvalue/2)
    )

  # breaks if less than 10 (in plot)
  jb_makebreaks <- function(
    xvalue
  ){
    minvalue <- min(xvalue, na.rm = T)
    maxvalue <- max(xvalue, na.rm = T)

    if (minvalue > 0.5) minvalue <- 0.49

    maxvalue <- ifelse(
      maxvalue < 3,
      3,
      ifelse(
        maxvalue < 5, 5,
        ifelse(
          maxvalue < 10, 10,
          maxvalue
        )
      )
    )

    xlimits <- c(minvalue,maxvalue)

    return(xlimits)
  }

  xbreaks <- c(0,0.5,1,1.5,2,5,10,15)




  # calculate the curves ------------------------------------------------
  ### If one estimate -------------------------------------------------
  if (is.na(est2.ll) | is.na(est2.ul)) {

    plot.data <- d.frame %>%
      dplyr::mutate(
        curve1 = ifelse(
          updown == "up",
          jb_getcurveup(
            jb_getrr(est1.ll,est1.ul),
            jb_getse(est1.ll,est1.ul,citype.pow),
            zvalue),
          jb_getcurvedown(
            jb_getrr(est1.ll,est1.ul),
            jb_getse(est1.ll,est1.ul,citype.pow),
            zvalue)
        )
      )

    # plot it
    xlimits <- jb_makebreaks(plot.data$curve1)

    plotout <- ggplot2::ggplot(plot.data,
                               ggplot2::aes_string('curve1', 'pvalue')
    ) +
      ggplot2::geom_line(size = functionwidth, colour = "#0033cc") +
      ggplot2::theme_bw() +
      ggplot2::theme(
        axis.text = ggplot2::element_text(size = labelsize),
        axis.title = ggplot2::element_text(size = labelsize)
      ) +
      ggplot2::ylab("p-value") +
      ggplot2::xlab(xlabel) +
      ggplot2::geom_vline(
        size = referencewidth,
        xintercept = 1,
        colour = "red",
        linetype = 2) +
      ggplot2::geom_hline(
        size = referencewidth,
        yintercept = citype.pval,
        colour = "red",
        linetype = 2) +
      ggplot2::scale_x_log10(
        breaks = xbreaks,
        limits = xlimits
      ) +
      ggplot2::scale_y_continuous(
        expand = c(0, 0), limits = c(0,1)
      )
  }

  ### If two estimates ---------------------------------------------------------------
  if (!is.na(est2.ll) | !is.na(est2.ul)) {
    d.frame <- d.frame %>%
      dplyr::mutate(
        curve1 = ifelse(
          updown == "up",
          jb_getcurveup(
            jb_getrr(est1.ll,est1.ul),
            jb_getse(est1.ll,est1.ul, citype.pow),
            zvalue),
          jb_getcurvedown(
            jb_getrr(est1.ll,est1.ul),
            jb_getse(est1.ll,est1.ul, citype.pow),
            zvalue)
        ),
        curve2 = ifelse(
          updown == "up",
          jb_getcurveup(
            jb_getrr(est2.ll,est2.ul),
            jb_getse(est2.ll,est2.ul, citype.pow),
            zvalue),
          jb_getcurvedown(
            jb_getrr(est2.ll,est2.ul),
            jb_getse(est2.ll,est2.ul, citype.pow),
            zvalue)
        )
      )

    # reshape data
    plot.data <- d.frame %>%
      dplyr::select(-c(zvalue,updown)) %>%
      tidyr::gather(curve,xvalue,-pvalue)

    # plot it
    xlimits <- jb_makebreaks(plot.data$xvalue)

    plotout <- ggplot2::ggplot(plot.data,
                               ggplot2::aes_string(
                                 'xvalue',
                                 'pvalue',
                                 colour = 'curve')
    ) +
      ggplot2::geom_line(size = functionwidth) +
      ggplot2::theme_bw() +
      ggplot2::theme(
        axis.text = ggplot2::element_text(size = labelsize),
        axis.title = ggplot2::element_text(size = labelsize),
        legend.text = ggplot2::element_text(size = labelsize)
      ) +
      ggplot2::ylab("p-value") +
      ggplot2::xlab(xlabel) +
      ggplot2::geom_vline(
        size = referencewidth,
        xintercept = 1,
        colour = "red",
        linetype = 2) +
      ggplot2::geom_hline(
        size = referencewidth,
        yintercept = citype.pval,
        colour = "red",
        linetype = 2) +
      ggplot2::scale_color_manual(
        name = "",
        labels = c(label1, label2),
        values = c("#0033cc", "#00cc00")
      ) +
      ggplot2::theme(legend.position = "top") +
      ggplot2::scale_x_log10(
        breaks = xbreaks,
        limits = xlimits
      ) +
      ggplot2::scale_y_continuous(
        expand = c(0, 0), limits = c(0,1)
      )
  }
  return(plotout)
}
